#region header
// Kay McCormick (mccor)
// 
// FileFinder3
// WpfApp1Tests3
// DuplicateKeyException.cs
// 
// 2020-01-20-6:26 AM
// 
// ---
#endregion

using System ;
using System.Runtime.Serialization ;

namespace WpfApp.Core.Exceptions
{
    /// <summary></summary>
    /// <seealso cref="System.Exception" />
    /// <autogeneratedoc />
    /// TODO Edit XML Comment Template for DuplicateKeyException
    [Serializable ]
    public class DuplicateKeyException : Exception
    {
        /// <summary>
        ///     Initializes a new instance of the <see cref="System.Exception" />
        ///     class.
        /// </summary>
        public DuplicateKeyException ( ) : this ( DefaultMessage ) { }

        /// <summary>
        ///     Initializes a new instance of the <see cref="System.Exception" />
        ///     class with a specified error message.
        /// </summary>
        /// <param name="message">The message that describes the error.</param>
        public DuplicateKeyException ( string message ) : base ( message ) { }

        /// <summary>
        ///     Initializes a new instance of the <see cref="System.Exception" />
        ///     class with a specified error message and a reference to the inner
        ///     exception
        ///     that is the cause of this exception.
        /// </summary>
        /// <param name="message">
        ///     The error message that explains the reason for the
        ///     exception.
        /// </param>
        /// <param name="innerException">
        ///     The exception that is the cause of the current
        ///     exception, or a null reference (<see langword="Nothing" /> in Visual
        ///     Basic)
        ///     if no inner exception is specified.
        /// </param>
        public DuplicateKeyException ( string message , Exception innerException ) : base (
                                                                                           message
                                                                                         , innerException
                                                                                          )
        {
        }

        /// <summary>
        ///     Initializes a new instance of the <see cref="System.Exception" />
        ///     class with a specified error message and a reference to the inner
        ///     exception
        ///     that is the cause of this exception.
        /// </summary>
        /// <param name="innerException">
        ///     The exception that is the cause of the current
        ///     exception, or a null reference (<see langword="Nothing" /> in Visual
        ///     Basic)
        ///     if no inner exception is specified.
        /// </param>
        public DuplicateKeyException ( Exception innerException ) : base (
                                                                          DefaultMessage
                                                                        , innerException
                                                                         )
        {
        }

        /// <summary>
        ///     Initializes a new instance of the <see cref="System.Exception" />
        ///     class with serialized data.
        /// </summary>
        /// <param name="info">
        ///     The
        ///     <see cref="System.Runtime.Serialization.SerializationInfo" /> that
        ///     holds
        ///     the serialized object data about the exception being thrown.
        /// </param>
        /// <param name="context">
        ///     The
        ///     <see cref="System.Runtime.Serialization.StreamingContext" /> that
        ///     contains contextual information about the source or destination.
        /// </param>
        /// <exception cref="T:System.ArgumentNullException">
        ///     <paramref name="info" /> is <see langword="null" />.
        /// </exception>
        /// <exception cref="T:System.Runtime.Serialization.SerializationException">
        ///     The
        ///     class name is <see langword="null" /> or
        ///     <see cref="P:System.Exception.HResult" /> is zero (0).
        /// </exception>
        protected DuplicateKeyException ( SerializationInfo info , StreamingContext context ) :
            base ( info , context )
        {
        }

        /// <summary>Initializes a new instance of the <see cref="DuplicateKeyException"/> class.</summary>
        /// <param name="key">The key.</param>
        /// <exception cref="System.ArgumentNullException">key</exception>
        /// <autogeneratedoc />
        /// TODO Edit XML Comment Template for #ctor
        public DuplicateKeyException ( object key ) : this ( $"{DefaultMessage}: {key}" )
        {
            Key = key ?? throw new ArgumentNullException ( nameof ( key ) ) ;
        }

        private static string DefaultMessage => "Duplicate Key" ;

        /// <summary>Gets the key.</summary>
        /// <value>The key.</value>
        /// <autogeneratedoc />
        /// TODO Edit XML Comment Template for Key
        public object Key { get ; }

        public override void GetObjectData(SerializationInfo info, StreamingContext context) => base.GetObjectData(info, context);
    }
}